### PR TITLE
Check appropriate OSSL_PARAM_get_* functions for NULL

### PR DESCRIPTION
**Pull request by nhorman**
The base type OSSL_PARAM getters will NULL deref if they are initalized as null.  Add NULL checks for those parameters that have no expectation of returning null (int32/64/uint32/64/BN).  Other types can be left as allowing NULL, as a NULL setting may be meaningful (string, utf8str, octet string, etc).

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

File: test/params_api_test.c [hunk 1]
```diff
@@ -157,6 +200,9 @@ static int test_param_int(int n)
                        sizeof(int) : raw_values[n].len;
     OSSL_PARAM param = OSSL_PARAM_int("a", NULL);
 
+    if (!TEST_int_eq(test_param_type_null(&param),1))
```
Comment (t8m) [inline test/params_api_test.c:203]: Nit: space after `,`

**Review by t8m — CHANGES_REQUESTED** [review]
One trivial nit

**Comment by tom-cosgrove-arm** [comment]
Is it reasonable to set `data` to `NULL` when `data_type` says it's an integer? If yes, this is okay. If not, then it seems reasonable for these to fault on null-pointer access.

**Comment by nhorman** [comment]
@tom-cosgrove-arm I believe so, yes.  We have a specific error condition that can be raised when the data type doesn't match the data stored in param (CRYPTO_R_PARAM_NOT_INTEGER_TYPE).  I could envision this being used to construct a parameter list of a given length that is filled out at a later time

File: test/params_api_test.c (hunk 1 above)
Comment (nhorman) [reply test/params_api_test.c]: fixed

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

File: crypto/params.c [hunk 2]
```diff
@@ -197,6 +197,10 @@ static int unsigned_from_unsigned(void *dest, size_t dest_len,
 /* General purpose get integer parameter call that handles odd sizes */
 static int general_get_int(const OSSL_PARAM *p, void *val, size_t val_size)
 {
+    if (p->data == NULL) {
+        err_not_integer;
+        return 0;
+    }
```
Comment (levitte) [inline crypto/params.c:203]: This should be `err_null_argument`

File: crypto/params.c [hunk 3]
```diff
@@ -226,6 +230,11 @@ static int general_set_int(OSSL_PARAM *p, void *val, size_t val_size)
 /* General purpose get unsigned integer parameter call that handles odd sizes */
 static int general_get_uint(const OSSL_PARAM *p, void *val, size_t val_size)
 {
+
+    if (p->data == NULL) {
+        err_not_integer;
+        return 0;
+    }
```
Comment (levitte) [inline crypto/params.c:237]: Should be `err_null_argument`

File: crypto/params.c [hunk 4]
```diff
@@ -385,6 +394,11 @@ int OSSL_PARAM_get_int32(const OSSL_PARAM *p, int32_t *val)
         return 0;
     }
 
+    if (p->data == NULL) {
+        err_not_integer;
+        return 0;
+    }
```
Comment (levitte) [inline crypto/params.c:400]: Should be `err_null_argument`

File: crypto/params.c [hunk 5]
```diff
@@ -534,6 +548,11 @@ int OSSL_PARAM_get_uint32(const OSSL_PARAM *p, uint32_t *val)
...
```
Comment (levitte) [inline crypto/params.c:554]: Should be `err_null_argument`

File: crypto/params.c [hunk 6]
```diff
@@ -685,6 +704,11 @@ int OSSL_PARAM_get_int64(const OSSL_PARAM *p, int64_t *val)
...
```
Comment (levitte) [inline crypto/params.c:710]: Should be `err_null_argument`

File: crypto/params.c [hunk 7]
```diff
@@ -829,6 +853,11 @@ int OSSL_PARAM_get_uint64(const OSSL_PARAM *p, uint64_t *val)
...
```
Comment (levitte) [inline crypto/params.c:859]: Should be `err_null_argument`

**Review by levitte — CHANGES_REQUESTED** [review]
Please be consistent.  The change of `OSSL_PARAM_get_BN` paves the way

**Comment by t8m** [comment]
This needs an update based on @levitte's review.

File: crypto/params.c (hunk 7 above)
Comment (nhorman) [reply crypto/params.c]: fixed

File: crypto/params.c (hunk 6 above)
Comment (nhorman) [reply crypto/params.c]: fixed

File: crypto/params.c (hunk 5 above)
Comment (nhorman) [reply crypto/params.c]: fixed

File: crypto/params.c (hunk 4 above)
Comment (nhorman) [reply crypto/params.c]: fixed

File: crypto/params.c (hunk 3 above)
Comment (nhorman) [reply crypto/params.c]: fixed

File: crypto/params.c (hunk 2 above)
Comment (nhorman) [reply crypto/params.c]: fixed

**Review by levitte — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master and 3.2 branches. This does not apply to 3.1 and 3.0 cleanly, please make a separate PR.

**Pull request closed without merging**